core/matrix/hyb.hpp

```diff
@@ -0,0 +1,401 @@
+     *     ell_stride is set to the number of rows of the matrix.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param partition  partition method
+     * @param val  the value used in partition (ignored in automatically)
+     */
+    Hyb(std::shared_ptr<const Executor> exec, const dim &size,
+        partition method = partition::automatically, index_type val = 0)
+        : Hyb(std::move(exec), size, size.num_cols, method, val)
+    {}
+
+    /**
+     * Creates an uninitialized Hyb matrix of the specified size and method.
+     *    (ell_stride is set to the number of rows of the matrix.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param max_nonzeros_per_row   maximum number of nonzeros in one row
+     * @param partition  partition method
+     * @param val  the value used in partition (ignored in automatically)
+     */
+    Hyb(std::shared_ptr<const Executor> exec, const dim &size,
+        size_type max_nonzeros_per_row,
+        partition method = partition::automatically, index_type val = 0)
+        : Hyb(std::move(exec), size, max_nonzeros_per_row, size.num_rows, {},
+              method, val)
+    {}
+
+    /**
+     * Creates an uninitialized Hyb matrix of the specified size and method.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param max_nonzeros_per_row   maximum number of nonzeros in one row
+     * @param stride                stride of the rows
+     * @param num_nonzeros  number of nonzeros
+     * @param partition  partition method
+     * @param val  the value used in partition (ignored in automatically)
+     */
+    Hyb(std::shared_ptr<const Executor> exec, const dim &size,
+        size_type max_nonzeros_per_row, size_type stride,
+        size_type num_nonzeros = {},
+        partition method = partition::automatically, index_type val = 0)
+        : EnableLinOp<Hyb>(exec, size),
+          method_(method),
+          method_val_(val),
+          ell_(std::move(
+              ell_type::create(exec, size, max_nonzeros_per_row, stride))),
+          coo_(std::move(coo_type::create(exec, size, num_nonzeros)))
+    {
+        check_method_val();
+    }
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    void check_method_val()
+    {
+        if (method_ == partition::automatically) {
+            method_val_ = 0;
+        } else if (method_ == partition::percent) {
+            method_val_ = std::min(static_cast<index_type>(100), method_val_);
+            method_val_ = std::max(zero<index_type>(), method_val_);
+        }
+    }
+
+private:
+    std::shared_ptr<ell_type> ell_;
+    std::shared_ptr<coo_type> coo_;
```

Comment:
I try the unique_ptr, but it is failed.
The problem seems to be in the convert_to function, because it wants to copy the unique pointer.
I get the output:
```
In file included from /home/u/mike/ginkgo/core/base/abstract_factory.hpp:38:0,
                 from /home/u/mike/ginkgo/core/base/lin_op.hpp:38,
                 from /home/u/mike/ginkgo/core/matrix/hyb.hpp:39,
                 from /home/u/mike/ginkgo/core/matrix/hyb.cpp:34:
/home/u/mike/ginkgo/core/base/polymorphic_object.hpp: In instantiation of ‘void gko::EnablePolymorphicAssignment<ConcreteType, ResultType>::convert_to(gko::EnablePolymorphicAssignment<ConcreteType, ResultType>::result_type*) const [with ConcreteType = gko::matrix::Hyb<std::complex<double>, long int>; ResultType = gko::matrix::Hyb<std::complex<double>, long int>; gko::EnablePolymorphicAssignment<ConcreteType, ResultType>::result_type = gko::matrix::Hyb<std::complex<double>, long int>]’:
/home/u/mike/ginkgo/core/matrix/hyb.cpp:249:1:   required from here
/home/u/mike/ginkgo/core/base/polymorphic_object.hpp:485:17: error: use of deleted function ‘gko::matrix::Hyb<std::complex<double>, long int>::Hyb(const gko::matrix::Hyb<std::complex<double>, long int>&)’
         *result = static_cast<result_type>(*self());
                 ^
In file included from /home/u/mike/ginkgo/core/matrix/hyb.cpp:34:0:
/home/u/mike/ginkgo/core/matrix/hyb.hpp:65:7: note: ‘gko::matrix::Hyb<std::complex<double>, long int>::Hyb(const gko::matrix::Hyb<std::complex<double>, long int>&)’ is implicitly deleted because the default definition would be ill-formed:
 class Hyb : public EnableLinOp<Hyb<ValueType, IndexType>>,
       ^
/home/u/mike/ginkgo/core/matrix/hyb.hpp:65:7: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = gko::matrix::Ell<std::complex<double>, long int>; _Dp = std::default_delete<gko::matrix::Ell<std::complex<double>, long int> >]’
In file included from /usr/include/c++/5/memory:81:0,
                 from /home/u/mike/ginkgo/core/base/array.hpp:38,
                 from /home/u/mike/ginkgo/core/matrix/hyb.hpp:38,
                 from /home/u/mike/ginkgo/core/matrix/hyb.cpp:34:
/usr/include/c++/5/bits/unique_ptr.h:359:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
```